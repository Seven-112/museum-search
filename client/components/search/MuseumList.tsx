import gql from "graphql-tag";
import { graphql } from "react-apollo";

/** MuseumList component props. */
interface MuseumListProps {
  query?: string;
}

/** MuseumList query response. */
interface MuseumListResponse {
  museums: any;
}

/** MuseumList query variables. */
interface MuseumListVariables {
  query?: string;
}

/** MuseumList query. */
const GET_MUSEUM_LIST = gql`
  query($query: String) {
    museums(query: $query, first: 50) {
      edges {
        node {
          id
          name
          streetAddress
          city
          state
          latitude
          longitude
        }
      }
      count
    }
  }
`;

/** Higher-order component to provide data from the GraphQL API to a React component. */
const withMuseumList = graphql<
  MuseumListProps,
  MuseumListResponse,
  MuseumListVariables
>(GET_MUSEUM_LIST, {
  options: ({ query }) => ({
    variables: {
      query
    }
  })
});

/** MuseumList component. */
export const MuseumList = withMuseumList(function MuseumList({
  data: { loading, error, museums }
}) {
  return (
    <ul className="list-group">
      {loading ? (
        "Loading..."
      ) : error ? (
        <div className="alert alert-danger">{error}</div>
      ) : museums ? (
        museums.edges.map(edge => (
          <li key={edge.node.id} className="list-group-item">
            <strong>{edge.node.name}</strong>
            <div>{edge.node.streetAddress}</div>
            <div>
              {[edge.node.city, edge.node.state]
                .filter(it => it || false)
                .join(" - ")}
            </div>
          </li>
        ))
      ) : null}
    </ul>
  );
});
