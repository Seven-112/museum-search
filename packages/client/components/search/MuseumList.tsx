import gql from "graphql-tag";
import { graphql } from "react-apollo";

/** MuseumList component props. */
interface IMuseumListProps {
  query?: string;
}

/** MuseumList query response. */
interface IMuseumListResponse {
  museums: any;
}

/** MuseumList query variables. */
interface IMuseumListVariables {
  query?: string;
}

/** MuseumList query. */
export const GET_MUSEUM_LIST = gql`
  query($query: String) {
    museums(query: $query, first: 50) {
      edges {
        node {
          id
          name
          streetAddress
          city
          state
          latitude
          longitude
        }
      }
      count
    }
  }
`;

/** Higher-order component to provide data from the GraphQL API to a React component. */
const withMuseumList = graphql<
  IMuseumListProps,
  IMuseumListResponse,
  IMuseumListVariables
>(GET_MUSEUM_LIST, {
  options: ({ query }) => ({
    variables: {
      query
    }
  })
});

/** MuseumList component. */
export const MuseumList = withMuseumList(function MuseumListInternal({
  data: { loading, error, museums }
}) {
  return (
    <ul className="list-group">
      {loading && "Loading..."}
      {error && <div className="alert alert-danger">{error.message}</div>}
      {museums &&
        museums.edges.map(edge => (
          <li key={edge.node.id} className="list-group-item">
            <strong>{edge.node.name}</strong>
            <div>{edge.node.streetAddress}</div>
            <div>
              {[edge.node.city, edge.node.state]
                .filter(it => it || false)
                .join(" - ")}
            </div>
          </li>
        ))}
    </ul>
  );
});
