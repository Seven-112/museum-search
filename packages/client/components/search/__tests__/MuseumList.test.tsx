import { mount } from "enzyme";
import toJson from "enzyme-to-json";
import { MockedProvider, MockedResponse } from "react-apollo/test-utils";
import { GET_MUSEUM_LIST, MuseumList } from "../MuseumList";
// tslint:disable-next-line: no-var-requires
const wait = require("waait");

const mocks: MockedResponse[] = [
  {
    request: {
      query: GET_MUSEUM_LIST,
      variables: {
        query: "space"
      }
    },
    result: {
      data: {
        museums: {
          __typename: "MuseumSearchConnection",
          count: 99,
          edges: [
            {
              __typename: "MuseumSearchEdge",
              cursor: "13911",
              node: {
                __typename: "Museum",
                city: null,
                id: 13911,
                latitude: 43.6555,
                longitude: -70.26151,
                name: "SPACE GALLERY",
                state: null,
                streetAddress: null
              }
            },
            {
              __typename: "MuseumSearchEdge",
              cursor: "14897",
              node: {
                __typename: "Museum",
                city: null,
                id: 14897,
                latitude: 42.26345,
                longitude: -85.60829,
                name: "SPACE GALLERY",
                state: null,
                streetAddress: null
              }
            },
            {
              __typename: "MuseumSearchEdge",
              cursor: "16357",
              node: {
                __typename: "Museum",
                city: "BONNE TERRE",
                id: 16357,
                latitude: 37.92174,
                longitude: -90.55071,
                name: "SPACE MUSEUM",
                state: "MO",
                streetAddress: "116 E SCHOOL ST"
              }
            }
          ]
        }
      }
    }
  },
  {
    error: new Error("An error happened!"),
    request: {
      query: GET_MUSEUM_LIST,
      variables: {
        query: "Give me an error!"
      }
    }
  }
];

describe("MuseumList component", () => {
  function mountWithContext(element: JSX.Element) {
    return mount(<MockedProvider mocks={mocks}>{element}</MockedProvider>);
  }

  it("Renders initially with a loading indicator.", () => {
    const wrapper = mountWithContext(<MuseumList query="space" />);
    expect(
      toJson(wrapper.find("MuseumListInternal").children())
    ).toMatchSnapshot();
  });

  it("Renders the museum list.", async () => {
    const wrapper = mountWithContext(<MuseumList query="space" />);
    await wait(0);
    wrapper.update();
    expect(
      toJson(wrapper.find("MuseumListInternal").children())
    ).toMatchSnapshot();
  });

  it("Renders an error.", async () => {
    const wrapper = mountWithContext(<MuseumList query="Give me an error!" />);
    await wait(2);
    wrapper.update();
    expect(
      toJson(wrapper.find("MuseumListInternal").children())
    ).toMatchSnapshot();
  });
});
