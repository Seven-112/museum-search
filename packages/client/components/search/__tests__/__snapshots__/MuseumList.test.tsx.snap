// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MuseumList component Renders an error. 1`] = `
<ul
  className="list-group"
>
  <div
    className="alert alert-danger"
  >
    Network error: An error happened!
  </div>
</ul>
`;

exports[`MuseumList component Renders initially with a loading indicator. 1`] = `
<ul
  className="list-group"
>
  Loading...
</ul>
`;

exports[`MuseumList component Renders the museum list. 1`] = `
<ul
  className="list-group"
>
  <li
    className="list-group-item"
    key="13911"
  >
    <strong>
      SPACE GALLERY
    </strong>
    <div />
    <div />
  </li>
  <li
    className="list-group-item"
    key="14897"
  >
    <strong>
      SPACE GALLERY
    </strong>
    <div />
    <div />
  </li>
  <li
    className="list-group-item"
    key="16357"
  >
    <strong>
      SPACE MUSEUM
    </strong>
    <div>
      116 E SCHOOL ST
    </div>
    <div>
      BONNE TERRE - MO
    </div>
  </li>
</ul>
`;
